blueprints/APIv3.tex

```diff
@@ -461,7 +478,79 @@ \subsection{Relationships as Objects}
+scheme.\\
+
+\subsection{JWT Authentication}
+Alternatively, authentication can be done using OAuth2 Bearer tokens in JSON Web-Token (JWT) format\footnote{The intention of this section
+is to declare that JWT-based authentication \emph{will} be supported by Traffic Ops API version 3 at some point. This does not necessarily
+mean it will be included in 3.0, but the standards defined here should be enough for clients to start building out support for it.},
+signed with JSON Web-Signatures (JWS). If a client requests a resource and is denied due to failed authentication, the endpoint MUST
+return a \code{401 Unauthorized} response code that contains an appropriate Alert as described in Section \ref{sec:401}, and also MUST set
+the "WWW-Authenticate" header to \emph{exactly} \code{Bearer realm="Apache Traffic Control"} by default. The realm may be configurable.\\
+Clients may present authentication in the "Authorization" HTTP header in the format \code{Bearer \emph{JWT}} where "JWT" is the
+base-64-encoded JWT (and accompanying JWS). When authentication using this method is unsuccessful, the endpoint MUST set the "WWW-Authenticate"
+HTTP header to include the "error" field set to "invalid\_token" and the "error\_description" field, which will likely have the same or similar
+content as the accompanying Alert. An example of such a failed response when the token presented by a client has expired is shown in Listing
+\ref{code:expired-token}.\newpage{}
+
+\begin{codelisting}
+\captionof{listing}{Expired Token Response}
+\label{code:expired-token}
+\begin{minted}[tabsize=2]{http}
+HTTP/1.1 401 Unauthorized
+Content-Type: application/json
+WWW-Authenticate: Bearer realm="Apache Traffic Control",
+                         error="invalid_token",
+                         error_description="token has expired"
+Transfer-Encoding: chunked
+
+{
+	"alerts": [
+		{
+			"level": "error",
+			"text": "Authentication failed; token has expired"
+		}
+	]
+}
+\end{minted}
+\end{codelisting}
+
+Valid JWTs consist of a JSON Object Signing and Encryption (JOSE) header followed by a JWS Payload and finally a JWS signature, as described
+in \href{https://tools.ietf.org/html/rfc7515}{RFC7515}. The JOSE header MUST define its "typ" field as "JWT" and its "alg" field as "HS256".
+This means that the JWS signature was generated using HMAC SHA-256 with the JWS Payload (un-encoded) as input. The JOSE header's "jwk" MUST be
+set to the value of the public key used to sign the SHA-256 hash in JSON Web Key (JWK) format as described in
+\href{https://tools.ietf.org/html/rfc7517}{RFC7517}.\\
+The claims of the JWS Payload - the actual JWT - MUST include the following members:
+
+\begin{itemize}
+	\item \code{iss} The issuer of the JWT. By default this should have the value of the network location of the Traffic Ops server that
+	generated the JWT - e.g. "trafficops.infra.ciab.test" - but may also refer to the LDAP server used when users authenticate using LDAP.
```

Comment:
The issuer field should always refer to the system which generated the JWT.  You can't get a JWT directly from ldap, so it should refer to the intermediary system that signed the JWT.